cheri.hpp

```diff
@@ -0,0 +1,98 @@
+#pragma once
+
+
+#if __has_feature(capabilities)
+#include "gc.hpp"
+#include <cheri/cheric.h>
+
+// On the stack inked list compression with cheri intrinsics.
+struct CheriCompressedListStyle
+{
+  static inline const uint32_t end = 0;
+
+  static const uint32_t transform(const void *ptr)
+  {
+    return static_cast<uint32_t>(cheri_getoffset(ptr));
+  }
+
+  static const void *get(uint32_t offset)
+  {
+    return cheri_setoffset(cheri_getstack(), offset);
+  }
+
+  static bool invalid(uint32_t offset)
+  {
+    return offset == 0;
+  }
+};
+
+template <> struct PointerSelector<CheriCompressedListStyle>
+{
+  using ptr_type = uint32_t;
+};
+
+// Stack Iterator that gives all the pointers it could find on the stack.
+struct CheriStackIterator
+{
+  void **location;
+
+  // Check if the pointer is pointing on the stack.
+  // Done by checking it aginst the bottom stack pointr's bounds.
+  bool is_stack(void *ptr)
+  {
+    return cheri_is_address_inbounds(bottom, cheri_getaddress(ptr));
+  }
+
+  // Check if the pointer is executable. 
+  // Done by using the permition bits.
+  bool is_exec(void *ptr)
+  {
+    return ((cheri_getperm(ptr) & 0b10) >> 1);
+  }
+
+  CheriStackIterator()
+  {
+    location = reinterpret_cast<void **>(bottom);
+  }
+
+  CheriStackIterator(void *start)
+  {
+    location = reinterpret_cast<void **>(start);
+  }
+
+  CheriStackIterator &operator++()
+  {
+    location -= 1;
+
+    // Skip all the things that arn't pointers.
```

Comment:
Aha, so I think I now understand what this function is doing. Let's document the function along the lines of:

```
// Move the iterator to the next pointer. Note that this will skip some intermediate words that can't
// possibly be pointers.
```